controller/instance_manager_controller.go

```diff
@@ -533,74 +568,106 @@ func (imc *InstanceManagerController) cleanUpPDBForNonExistingIM() error {
 }
 
 func (imc *InstanceManagerController) deleteInstanceManagerPDB(im *longhorn.InstanceManager) error {
-	err := imc.ds.DeletePDB(imc.getPDBName(im))
+	name := imc.getPDBName(im)
+	err := imc.ds.DeletePDB(name)
 	if err != nil && !datastore.ErrorIsNotFound(err) {
 		return err
 	}
+	imc.logger.Infof("Deleted %v PDB", name)
 	return nil
 }
 
-func (imc *InstanceManagerController) canDeleteInstanceManagerPDB(im *longhorn.InstanceManager) (bool, error) {
+func (imc *InstanceManagerController) canDeleteInstanceManagerPDB(im *longhorn.InstanceManager) (bool, string, error) {
 	// If there is no engine instance process inside the engine instance manager,
 	// it means that all volumes are detached.
 	// We can delete the PodDisruptionBudget for the engine instance manager.
 	if im.Spec.Type == longhorn.InstanceManagerTypeEngine {
 		if len(im.Status.Instances) == 0 {
-			return true, nil
+			return true, "", nil
 		}
-		return false, nil
+		return false, "", nil
 	}
 
 	// Make sure that the instance manager is of type replica
 	if im.Spec.Type != longhorn.InstanceManagerTypeReplica {
-		return false, fmt.Errorf("the instance manager %v has invalid type: %v ", im.Name, im.Spec.Type)
+		return false, "", fmt.Errorf("the instance manager %v has invalid type: %v ", im.Name, im.Spec.Type)
 	}
 
-	// Must wait for all volumes detached from the current node first.
-	// This also means that we must wait until the PDB of engine instance manager on the current node is deleted
-	allVolumeDetached, err := imc.areAllVolumesDetachedFromCurrentNode()
+	// Must wait for all instance removed from the current node first.
+	allInstanceRemoved, err := imc.areAllInstanceRemovedFromCurrentNode(im)
 	if err != nil {
-		return false, err
+		return false, "", err
 	}
-	if !allVolumeDetached {
-		return false, nil
+	if !allInstanceRemoved {
+		return false, "", nil
 	}
```

Comment:
Wait... Can we replace the volume detachment check function with the instance cleanup wait function?
After this modification, can a node with running replicas be drained even if the replicas are not the only ones of the volumes?

Reply:
Reverted.